arangod/Aql/CollectNode.cpp

```diff
@@ -437,43 +442,186 @@ ExecutionNode* CollectNode::clone(ExecutionPlan* plan, bool withDependencies,
+    case ExecutionNode::COLLECT:
+    case ExecutionNode::SCATTER:
+    case ExecutionNode::GATHER:
+    case ExecutionNode::REMOTE:
+    case ExecutionNode::INSERT:
+    case ExecutionNode::REMOVE:
+    case ExecutionNode::REPLACE:
+    case ExecutionNode::UPDATE:
+    case ExecutionNode::RETURN:
+    case ExecutionNode::NORESULTS:
+    case ExecutionNode::DISTRIBUTE:
+    case ExecutionNode::UPSERT:
+    case ExecutionNode::TRAVERSAL:
+    case ExecutionNode::INDEX:
+    case ExecutionNode::SHORTEST_PATH:
+    case ExecutionNode::K_SHORTEST_PATHS:
+    case ExecutionNode::REMOTESINGLE:
+    case ExecutionNode::ENUMERATE_IRESEARCH_VIEW:
+    case ExecutionNode::DISTRIBUTE_CONSUMER:
+    case ExecutionNode::SUBQUERY_END:
+    case ExecutionNode::MATERIALIZE:
+      return false;
+    case ExecutionNode::MAX_NODE_TYPE_VALUE:
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL_AQL);
+  }
+  TRI_ASSERT(false);
+  THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL_AQL);
+}
 
-  ~UserVarFinder() = default;
+auto isVariableInvalidatingNode(ExecutionNode const& node) -> bool {
+  switch (node.getType()) {
+    case ExecutionNode::SINGLETON:
+    case ExecutionNode::SUBQUERY_START:
+    case ExecutionNode::COLLECT:
+      return true;
+    case ExecutionNode::ENUMERATE_COLLECTION:
+    case ExecutionNode::ENUMERATE_LIST:
+    case ExecutionNode::FILTER:
+    case ExecutionNode::LIMIT:
+    case ExecutionNode::CALCULATION:
+    case ExecutionNode::SUBQUERY:
+    case ExecutionNode::SORT:
+    case ExecutionNode::SCATTER:
+    case ExecutionNode::GATHER:
+    case ExecutionNode::REMOTE:
+    case ExecutionNode::INSERT:
+    case ExecutionNode::REMOVE:
+    case ExecutionNode::REPLACE:
+    case ExecutionNode::UPDATE:
+    case ExecutionNode::RETURN:
+    case ExecutionNode::NORESULTS:
+    case ExecutionNode::DISTRIBUTE:
+    case ExecutionNode::UPSERT:
+    case ExecutionNode::TRAVERSAL:
+    case ExecutionNode::INDEX:
+    case ExecutionNode::SHORTEST_PATH:
+    case ExecutionNode::K_SHORTEST_PATHS:
+    case ExecutionNode::REMOTESINGLE:
+    case ExecutionNode::ENUMERATE_IRESEARCH_VIEW:
+    case ExecutionNode::DISTRIBUTE_CONSUMER:
+    case ExecutionNode::SUBQUERY_END:
+    case ExecutionNode::MATERIALIZE:
+      return false;
+    case ExecutionNode::MAX_NODE_TYPE_VALUE:
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL_AQL);
+  }
+  TRI_ASSERT(false);
+  THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL_AQL);
+}
 
-  std::vector<Variable const*> userVars;
-  int mindepth;  // minimal depth to consider
-  int depth;
+auto isLoop(ExecutionNode const& node) -> bool {
+  switch (node.getType()) {
+    case ExecutionNode::ENUMERATE_COLLECTION:
+    case ExecutionNode::INDEX:
+    case ExecutionNode::ENUMERATE_LIST:
+    case ExecutionNode::TRAVERSAL:
+    case ExecutionNode::SHORTEST_PATH:
+    case ExecutionNode::K_SHORTEST_PATHS:
+    case ExecutionNode::ENUMERATE_IRESEARCH_VIEW:
+    case ExecutionNode::COLLECT:
+      // TODO must there be more? e.g. FILTER?
```

Comment:
Good point, this TODO is actually open. While this is called "isLoop", actually the question is what qualifies as "top-level" here. As in a query
```
LET foo = ...
SOME_OP
COLLECT ...
...
```
Now does SOME_OP let COLLECT stay on the "top-level" or not. In this case it affects whether the variable would be collected or not. But it also, and more importantly, affects which variables qualify as "top-level":
```
SOME_OP
LET foo = ...
...
FOR x IN ...
...
COLLECT ... INTO g
```
Now, if SOME_OP is not included in that list, `foo` still qualifies as top-level. That means COLLECT will *not* collect it, because the variable is assumed to be usable anyway after that - while the latter is actually decided in the parser.

That's why I thought at least `FILTER` should be included (and the function then renamed).

Reply:
As discussed, this is fine.